### DISCUSSION TITLE
libbpf tc program

### DISCUSSION THREAD

**Discussion by cheneytianx** [libbpf/libbpf-bootstrap#101] [Q&A] [answered]
How to attach tc program in the userspace code using libbpf?

**Comment by vincentmli** [comment]
a quick google search found this https://lore.kernel.org/lkml/20210402152743.dbadpgcmrgjt4eca@apollo/T/#m773e6e57e9c763eb9590c86304a0473a8b187f23

**Comment by chenhengqi** [comment, marked as answer]
```c
DECLARE_LIBBPF_OPTS(bpf_tc_hook, tc_hook,
	.ifindex = 4, .attach_point = BPF_TC_EGRESS);
DECLARE_LIBBPF_OPTS(bpf_tc_opts, tc_opts,
	.flags = BPF_TC_F_REPLACE, .handle = 1, .priority = 1);

err = bpf_tc_hook_create(&tc_hook);
if (err)
	return err;

tc_opts.prog_fd = bpf_program__fd(skel->progs.tc_egress);
err = bpf_tc_attach(&tc_hook, &tc_opts);
if (err)
	return err;
```

**Reply by cheneytianx** [reply]
Thanks for your help! 
Any references for the parameters in `DECLARE_LIBBPF_OPTS` that are related to tc , like `handle` and `priority` ? 
What are the meaning of them and can we set different values to them?

**Reply by vincentmli** [reply]
@chenhengqi since I did not see `direct action` in your suggestion, is that set by default ?

**Reply by chenhengqi** [reply]
@vincentmli  Please refer to https://github.com/libbpf/libbpf/blob/7436656dbf6f07f5eb4a5a5f2f9ef5a13e5ff9db/src/netlink.c#L668
